packages/vm/src/environment.rs

```diff
@@ -372,6 +376,66 @@ impl<A: BackendApi, S: Storage, Q: Querier> Environment<A, S, Q> {
             None => Err(VmError::uninitialized_context_data("callee_func_metadata")),
         }
     }
+
+    pub fn try_record_dynamic_call_trace(&self) -> VmResult<()> {
+        self.with_context_data_mut(|ctx| {
+            if ctx.dynamic_callstack.len() >= DYNAMIC_CALL_DEPTH_LIMIT_CNT {
+                return Err(VmError::dynamic_call_depth_over_limitation_err());
+            }
+
+            let contract_env: Env = match &ctx.serialized_env {
+                Some(env) => from_slice(&env),
+                None => Err(VmError::uninitialized_context_data("serialized_env")),
+            }?;
+            match ctx
+                .dynamic_callstack
+                .iter()
+                .find(|x| **x == contract_env.contract.address)
+            {
+                Some(_) => Err(VmError::re_entrancy_err()),
+                None => {
+                    ctx.dynamic_callstack.push(contract_env.contract.address);
+                    Ok(())
+                }
+            }
+        })
+    }
+
+    pub fn remove_latest_dynamic_call_trace(&self) {
+        self.with_context_data_mut(|ctx| {
+            ctx.dynamic_callstack.pop();
+        })
+    }
+
+    //pass_callstack will be called through wasmvm.
+    pub fn pass_callstack<A2, S2, Q2>(&self, target: &mut Environment<A2, S2, Q2>) -> VmResult<()>
```

Comment:
It is for not only passing but also checking the reentrancy. Function name should explain it (about checking).